### PR TITLE
Simplify `concat_ws(null, ..)` to `null`

### LINKED ISSUE #3607: Simplify `CONCAT_WS(NULL, ..)` to `NULL`
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**
This is a subtask of #3599.
When the separator in the `concat_ws` expression is `null`, the value is always null.
So we can do this simplification.

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by HaoYang670**
Signed-off-by: remzi <13716567376yh@gmail.com>

# Which issue does this PR close?

Closes #3607.

 # Rationale for this change

## Before the optimization
```sql
❯ create table t as select 'aa' as a, 'bb' as b, 'cc' as c;
0 rows in set. Query took 0.008 seconds.
❯ explain select concat_ws(null, a, b, c) from t;
+---------------+------------------------------------------------------------------------------------------------------------------------+
| plan_type     | plan                                                                                                                   |
+---------------+------------------------------------------------------------------------------------------------------------------------+
| logical_plan  | Projection: concatwithseparator(NULL, #t.a, #t.b, #t.c)                                                                |
|               |   TableScan: t projection=[a, b, c]                                                                                    |
| physical_plan | ProjectionExec: expr=[concatwithseparator(CAST(NULL AS Utf8), a@0, b@1, c@2) as concatwithseparator(NULL,t.a,t.b,t.c)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                                                                        |
|               |                                                                                                                        |
+---------------+------------------------------------------------------------------------------------------------------------------------+
```

## After the optimization
```sql
❯ create table t as select 'aa' as a, 'bb' as b, 'cc' as c;
0 rows in set. Query took 0.009 seconds.
❯ explain select concat_ws(null, a, b, c) from t;
+---------------+----------------------------------------------------------------------+
| plan_type     | plan                                                                 |
+---------------+----------------------------------------------------------------------+
| logical_plan  | Projection: Utf8(NULL) AS concatwithseparator(NULL,t.a,t.b,t.c)      |
|               |   TableScan: t projection=[a]                                        |
| physical_plan | ProjectionExec: expr=[NULL as concatwithseparator(NULL,t.a,t.b,t.c)] |
|               |   MemoryExec: partitions=1, partition_sizes=[1]                      |
|               |                                                                      |
+---------------+----------------------------------------------------------------------+
```

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 1]
```diff
@@ -816,6 +815,23 @@ impl<'a, S: SimplifyInfo> ExprRewriter for Simplifier<'a, S> {
                 out_expr.rewrite(self)?
             }
 
+            // concat_ws
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:818]:
```suggestion
            // concat_ws
            // If any of the arguments is NULL, the output will be null
```

File: datafusion/optimizer/src/simplify_expressions.rs [hunk 2]
```diff
@@ -1132,6 +1148,40 @@ mod tests {
...
+            }
+        }
+
+        let null = Expr::Literal(ScalarValue::Utf8(None));
+        // simple test
+        {
+            let expr = build_concat_ws_expr(&[null.clone(), col("c1"), col("c2")]);
+            assert_eq!(simplify(expr), null);
+        }
+
```
Comment (alamb) [inline datafusion/optimizer/src/simplify_expressions.rs:1166]: I wonder if we could support when arguments other than the first are null? For example

```suggestion
        // simple test
        {
            let expr = build_concat_ws_expr(&[null.clone(), col("c1"), col("c2")]);
            assert_eq!(simplify(expr), null);
        }
        // simple test with null further down the argument list
        {
            let expr = build_concat_ws_expr(&[col("c1"), null.clone(), col("c2")]);
            assert_eq!(simplify(expr), null);
        }
```

I think the current code may only detect NULL in the first argument position

**Review by alamb — APPROVED** [review]
Thank you @HaoYang670  -- this change makes sense to me and I think it could be merged as is. However, I think it is worth considering handling nulls in positions other than the first as well

**Comment by HaoYang670** [comment]
> However, I think it is worth considering handling nulls in positions other than the first as well

@alamb, if the separator is `null`, the `concat_ws` expression will always have the value `null`,

If `null` appears in other positions, then the `null` value will be ignored:
```
concat_ws(" | ", col1, null, col2) --> concat_ws("|", col1, col2)
```

So we should only check `null` at the first position

**Comment by alamb** [comment]
> So we should only check null at the first position

Perhaps we can add a test for that behavior explicitly (as in a test that a null at the second position does not result in a rewrite) to make it clear that is the expected behavior

**Comment by HaoYang670** [comment]
Will add.

**Comment by alamb** [comment]
In case anyone else is interested in how `concat_ws` works, I tested it out with datafusion-cli

```sql
❯ select concat_ws('--', 'foo', 'bar', 'baz');
+---------------------------------------------------------------------+
| concatwithseparator(Utf8("--"),Utf8("foo"),Utf8("bar"),Utf8("baz")) |
+---------------------------------------------------------------------+
| foo--bar--baz                                                       |
+---------------------------------------------------------------------+
1 row in set. Query took 0.000 seconds.
❯ select concat_ws('--', null, 'bar', 'baz');
+--------------------------------------------------------------+
| concatwithseparator(Utf8("--"),NULL,Utf8("bar"),Utf8("baz")) |
+--------------------------------------------------------------+
| bar--baz                                                     |
+--------------------------------------------------------------+
1 row in set. Query took 0.003 seconds.
❯ select concat_ws(null, 'foo', 'bar', 'baz');
+---------------------------------------------------------------+
| concatwithseparator(NULL,Utf8("foo"),Utf8("bar"),Utf8("baz")) |
+---------------------------------------------------------------+
|                                                               |
+---------------------------------------------------------------+
1 row in set. Query took 0.000 seconds.
```

**Review by alamb — APPROVED** [review]
LGTM -- thanks @HaoYang670
